### ISSUE TITLE
Optimization of docs for non-Yocto build systems

### ISSUE THREAD

**Issue by fkromer** [mendersoftware/mender-docs#79]
The [Mender docs section about device integration with other build systems than Yocto](https://docs.mender.io/Devices/System-requirements#other-build-systems) references some [post about porting Mender to a non-Yocto build system](https://mender.io/blog/porting-mender-to-a-non-yocto-build-system). Unfortunately the post for non-Yocto build systems does reference the docs for Yocto build systems a lot and uses terminology related to files which are generated or patched by Yocto which is not obvious for users which do not use Yocto a lot. As a result the procedure for non-Yocto build systems is hard to follow.

What would be the best approach to add the non-Yocto procedure into the docs (little duplication, etc.)?

**Comment by eysteins** [comment]
We are focusing on supporting Yocto now in the beginning, and looking to expand later as there is demand for it. The reason we did not make this part of the official docs is that we are unable to maintain it at this stage; it is much easier to maintain the Yoco scripts as they can be readily tested with CI.

However, we would be happy to assist with community contributions on this as long as we can distinguish it from the "officially supported" docs which we maintain and test all the time.

To your question, I am not sure what the best approach is, perhaps @kacf can chime in. Are you able to assist with this as well @fkromer if we make a page in the docs site?

**Comment by fkromer** [comment]
@eysteins :heavy_check_mark: Sure. Working on a more detailed generic migration/build procedure documentation helps to speed up everyones specifc development as well.

**Comment by kacf** [comment]
My blog post would inherently be Yocto-centric, since as a writer I am biased, and it serves more as an example than actual documentation. I think if you (or someone else in the community) have enough expertise with a different build system to write the sections you mention, then I'm all for it.

**Comment by eysteins** [comment]
@fkromer thanks for the feedback and interest in contributing.

The current docs make assumptions that Yocto is being used throughout, and this helps make it easier for users of Yocto. If we add information about non-Yocto as well without thinking through the documentation architecture I think it will be a bit messy and not serve either audience well.

I like your idea of tailoring it to BeageBone official images, as that aligns well with the rest of the docs.

So for now, what I would propose is a separate section of the docs called something like "Generic device integration", perhaps below "Client configuration". In there, we could have subsections on the what needs to be in place like outlined in the [blog post](https://mender.io/blog/porting-mender-to-a-non-yocto-build-system) as well as tutorials for enabling Mender OTA using official BeagleBone images.

Does that make sense? We'd be happy to review any PRs you make in this regard @fkromer !

**Comment by fkromer** [comment]
@eysteins A description of how to get to an *eMMC image with the partitioning required by Mender* (a generic description and a BeagleBone Black specific description) instead of the [SD card disk image (mender-beaglebone.sdimg.gz) with the partitioning required by Mender](https://docs.mender.io/Getting-started/Deploy-to-physical-devices#prepare-the-disk-image), the differences to be considered in the [modification of the (SD card) image](https://docs.mender.io/Artifacts/Modifying-a-disk-image) and in the [u-boot integration](https://docs.mender.io/Devices/Integrating-with-U-Boot) would fit in a "generic device integration" section well. That should close the gap, right? I still worry about required cross-referencing then...

**Comment by eysteins** [comment]
Yes, I think so.

We can always move things around and adjust links once we have the docs, I just think it would be easier to start off with a fairly isolated docs section for non-Yocto users. It will receive more attention and become more integrated going forward if people find it useful for sure. :)

**Comment by fkromer** [comment]
I started the docs for the system image partitioning in [my fork (master branch)](https://github.com/fkromer/mender-docs/blob/master/06.Generic-device-integration/02.System-image-partitioning/docs.md). But the rootfs partition of the used example image "Recommended Debian Images / Jessie for BeagleBone via microSD card / Debian 8.6 2016-11-06 4GB SD LXQ" from the [BeagleBone image download site](https://beagleboard.org/latest-images) is to big (~3GB) to put the 2 other partitions (rootfs2, data) on it. Is there a way to shrink this partition somehow? (a very quick web search did match...) Otherwise I would switch to an older image with a size of ~ 2GB like "Older Debian images / BeagleBone Black (eMMC flasher) / Debian 7.5 (BeagleBone Black - 2GB eMMC) 2014-05-14" (https://debian.beagleboard.org/images/BBB-eMMC-flasher-debian-7.5-2014-05-14-2gb.img.xz).

**Comment by estenberg** [comment]
Nice, this looks interesting!

I don't know of a way do shrink the rootfs in general except for compression or removing contents.

Regarding compression it might be possible to convert it into a squashfs; this is a bit experimental though as I don't know of Mender users that use squashfs. It should be possible as it operates on block devices. The other question is how the beaglebone tools will behave with a readonly file system, probably it won't "just" work, so one have to combine it with a e.g. unionfs, overlayfs, aufs to store the writes.

Removing some things may work, by following your instructions I see that the image is about 2.8GB on disk, so we should ideally shave off about 1 GB. There are some rather large directories that I'm not sure are important or not:

```
114596	/mnt/rootfs/usr/share/doc
151932	/mnt/rootfs/usr/share/locale
210736	/mnt/rootfs/usr/share/ti
```
These three alone are close to 500 MB.

So with some research and experimentation it should be possible to get this to work.

The final option is to get a SD card for this for now, with at least 8GB size. :-)

**Comment by fkromer** [comment]
@estenberg 
> ... we should ideally shave off about 1 GB

I think so, because just Revision C of the BeagleBone Black seems to have a eMMC size of 4 GB.

> The final option is to get a SD card for this for now, with at least 8GB size. :-)

I want to avoid that because using a SD card is not really a "production" like use case ;-)

**Comment by fkromer** [comment]
I added an introductional [section about a Yocto vs. generic build system comparison](https://github.com/fkromer/mender-docs/blob/master/07.Generic-device-integration/03.Build-system-overview/docs.md). Feel free to review and pull request there as well...

**Comment by estenberg** [comment]
Thanks! I will take a deeper look shortly!

**Comment by estenberg** [comment]
The new section looks good for explaining the current build process.
I am a little concerned that it will quickly grow out of date as the meta layer changes, though.

Maybe we should focus on making the scripts and steps required to "convert" the official BBB image first? If it is possible then that might be a reusable method for other builds as well.

**Comment by fkromer** [comment]
@estenberg I totaly agree. In the end it would probably be the best to just have some instructions and directory/file patches (~ `diff`) related to the official image which can be explained in the docs shortly. These patches and explanations can be easily adjsuted to new versions of the official image of the BeagleBone Black and changes regarding the integration with Mender.

But I am still stuck regarding the shrinking of the official 4 GB sd card image to get 2 rootfs onto the full image. Maybee I will just use the official older 2 GB image to get to the Mender compatible full image...

**Comment by estenberg** [comment]
Sounds good. Yeah, I don't have much input on shrinking the image except the few ideas above, unfortunately. That said, we will probably look into squashfs some time in the future, so if you (or other community members) are interested in that we can certainly assist. :)

**Comment by fkromer** [comment]
@kacf Your [non-Yocto post](https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc) references a [.inc file](https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc) which adds `/include/config_mender_defines.h` to the u-boot sources. I was not able to figure out how

- `${MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1}` for `#define MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1 ${MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1}` and

- `${MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_2}` for `#define MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_2 ${MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_2}`

are replaced during the Yocto build. How do I figure out how to set these?

**Comment by kacf** [comment]
You mean which value they are set to? It's complicated I'll admit. They are set [here](https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc#L74-L87), but these values are also influenced by other values. If you grep for each variable name in meta-mender it should be possible to figure out where they are set.

**Comment by fkromer** [comment]
@kacf Thanks for that hint.

**Comment by fkromer** [comment]
I wrote a short Python script to check if I understood the calculation prodecure:

    #!/usr/bin/env python

    # defined in /meta-mender/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc
    def mender_get_env_total_aligned_size(bootenv_size, alignment_mb):
        alignment_bytes = alignment_mb * 1024 * 1024
        env_aligned_size = int((bootenv_size + alignment_bytes - 1) / alignment_bytes) * alignment_bytes

        # Total size, original and redundant environment.
        total_env_size = env_aligned_size * 2

        return "%d" % total_env_size

    # defined in /meta-mender/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc
    def mender_get_env_offset(start_offset, index, total_aligned_size):
        if index == 1:
            return "0x%x" % int(start_offset)
        elif index == 2:
            return "0x%x" % int(start_offset + total_aligned_size / 2)
        else:
            raise Exception("env index out of range in mender_get_env_offset: Should not happen")

    # defined in /meta-mender/meta-mender-core/classes/mender-uboot.bbclass
    def mender_mb2bytes(mb):
        return mb * 1024 * 1024

    if __name__ == "__main__":
        # defined in /meta-mender/meta-mender-core/classes/mender-install.bbclass
        MENDER_PARTITION_ALIGNMENT_MB = 8

        # defined in /meta-mender/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc
        BOOTENV_SIZE = 0x20000

        # /meta-mender/meta-mender-core/classes/mender-uboot.bbclass
        MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET = mender_mb2bytes(MENDER_PARTITION_ALIGNMENT_MB)

        # /meta-mender/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc
        MENDER_BOOTENV_TOTAL_ALIGNED_SIZE = mender_get_env_total_aligned_size(BOOTENV_SIZE, MENDER_PARTITION_ALIGNMENT_MB)

        # /meta-mender/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc
        MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1 = mender_get_env_offset(MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET, 1, MENDER_BOOTENV_TOTAL_ALIGNED_SIZE)
        MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_2 = mender_get_env_offset(MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET, 2, MENDER_BOOTENV_TOTAL_ALIGNED_SIZE)

But it seems like I missed something... when I run the script (here `mender.py`) I get an error message displayed:

    ~/ws_gitlab$ python ./mender.py
    Traceback (most recent call last):
      File "./mender.py", line 41, in <module>
        MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_2 = mender_get_env_offset(MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET, 2, MENDER_BOOTENV_TOTAL_ALIGNED_SIZE)
      File "./mender.py", line 18, in mender_get_env_offset
        return "0x%x" % int(start_offset + total_aligned_size / 2)
    TypeError: unsupported operand type(s) for /: 'str' and 'int'

**Comment by kacf** [comment]
Change this:

```
@@ -8,7 +8,7 @@
     # Total size, original and redundant environment.
     total_env_size = env_aligned_size * 2
 
-    return "%d" % total_env_size
+    return total_env_size
 
 # defined in /meta-mender/meta-mender-core/recipes-bsp/u-boot/u-boot-mender-common.inc
 def mender_get_env_offset(start_offset, index, total_aligned_size):
```

The reason it's there is just because bitbake does some additional string magic behind the scenes.

**Comment by fkromer** [comment]
Thanks for that "behind the scenes" hint :thumbsup:

**Comment by fkromer** [comment]
Is it normal that the values for `MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1` and `MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1` do not change if I use `BOOTENV_SIZE = 0x2000` instead of `BOOTENV_SIZE = 0x20000`?

**Comment by kacf** [comment]
Yes, because they are aligned to `MENDER_PARTITION_ALIGNMENT_MB`.

**Comment by fkromer** [comment]
Great. (I just wanted to make sure not to miss some other "hidden magic".)

**Comment by fkromer** [comment]
I have built a custom u-boot (not for the BeagleBone because I am still blocked due to the image size...) with the modifications for the Mender integration. When I do not abort the boot process it hangs up after displaying the debug message`[    5.254455] Waiting for root device ${mender_kernel_root}...`. The mender specific u-boot environment variables are as follows:

    bootcount=1
    bootdelay=1
    bootlimit=1
    mender_altbootcmd=if test ${mender_boot_part} = 2; then setenv mender_boot_part 3; else setenv
    mender_boot_part 2; fi; setenv upgrade_available 0; saveenv; run mender_setup
    mender_boot_part=2
    mender_setup=setenv mender_kernel_root /dev/mmcblk0p${mender_boot_part}; setenv mender_uboot_root mmc 0:${mender_boot_part}; setenv expand_bootargs setenv bootargs ${bootargs}; run expand_bootargs; setenv expand_bootargs
    mender_try_to_recover=if test ${upgrade_available} = 1; then reset; fi
    mender_uboot_boot=mmc 0:1
    mender_uboot_dev=0
    mender_uboot_if=mmc
    upgrade_available=0

The value `/dev/mmcblk0p2` for `mender_kernel_root` should be ok. What else could be missing/wrong?

**Comment by kacf** [comment]
Does your `bootcmd` call `run mender_setup` as its first step? This is required.

**Comment by fkromer** [comment]
Yes, it does. (I added the `bootcmd` above.)

**Comment by kacf** [comment]
Uhm... Are you sure? I don't see it.

**Comment by fkromer** [comment]
Oh, sorry... now its added (multitasking can be bad).

**Comment by kacf** [comment]
Hmm, it looks correct as far as I can see. Maybe you can step through each of the commands in `bootcmd` manually, and verify that `mender_kernel_root` and `bootargs` are set correctly, using `printenv`.

**Comment by fkromer** [comment]
That was a good hint: When executing the first command in `bootcmd` the `mender_kernel_root` and `mender_uboot_root` seem to be set correctly. (The `bootargs` are not set yet.)

    Apalis TK1 # run mender_setup
    Apalis TK1 # printenv mender_kernel_root
    mender_kernel_root=/dev/mmcblk0p2
    Apalis TK1 # printenv mender_uboot_root
    mender_uboot_root=mmc 0:2
    Apalis TK1 # printenv expand_bootargs
    ## Error: "expand_bootargs" not defined
    Apalis TK1 # printenv bootargs
    ## Error: "bootargs" not defined

When executing the single commands in `emmcboot` before the dtb and kernel is loaded the `mender_kernel_root` seems to be set not correctly for some reason...:

    Apalis TK1 # run setup                                                                                                                                                                                                                                                                                       
    Apalis TK1 # setenv bootargs ${defargs} ${emmcargs} ${setupargs} ${vidargs}
    Apalis TK1 # printenv bootargs
    bootargs=lp0_vec=2064@0xf46ff000 core_edp_mv=1150 core_edp_ma=4000 usb_port_owner_info=2 lane_owner_info=6 emc_max_dvfs=0 ip=off root=${mender_kernel_root} rw rootfstype=ext3 rootwait igb_mac=00:14:2d:2b:b5:a3 consoleblank=0 no_console_suspend=1 console=tty1 console=ttyS0,115200n8 debug_uartport=lsport,0 video=tegrafb0:640x480-16@60 fbcon=map:1
    Apalis TK1 # printenv mender_kernel_root
    mender_kernel_root=/dev/mmcblk0p2

Is there some dependency of `mender_setup` that `bootargs` needs to be set prior to its call?

**Comment by kacf** [comment]
> Is there some dependency of mender_setup that bootargs needs to be set prior to its call?

Yes, probably they should be swapped, so that `setup` is called first.

**Comment by fkromer** [comment]
@kacf Thanks a lot for the help. Seems like the `bootargs` need to be set prior to calling `run mender_setup;` because it works fine with

    bootcmd=run emmcboot; run mender_try_to_recover; setenv fdtfile ${soc}-apalis-${fdt_board}.dtb && run distro_bootcmd
    emmcboot=run setup; setenv bootargs ${defargs} ${emmcargs} ${setupargs} ${vidargs}; run mender_setup; echo Booting from internal eMMC chip...; run emmcdtbload; load ${mender_uboot_root} ${kernel_addr_r} ${boot_file} && run fdt_fixup && bootm ${kernel_addr_r} - ${dtbparam}

instead of

    bootcmd=run mender_setup; run emmcboot; run mender_try_to_recover; setenv fdtfile ${soc}-apalis-${fdt_board}.dtb && run distro_bootcmd
    emmcboot=run setup; setenv bootargs ${defargs} ${emmcargs} ${setupargs} ${vidargs}; echo Booting from internal eMMC chip...; run emmcdtbload; load ${mender_uboot_root} ${kernel_addr_r} ${boot_file} && run fdt_fixup && bootm ${kernel_addr_r} - ${dtbparam}

**Comment by kacf** [comment]
No problem, glad it worked out! :-)

**Comment by fkromer** [comment]
I need to place into the **rootfs A /etc/fstab** this
```
/dev/mmcblk0p1   /uboot               vfat       defaults,sync,auto    0  0
/dev/mmcblk0p2   /                    ext3       defaults              0  0
/dev/mmcblk0p4   /data                ext3       defaults              0  0
```
and into the **rootfs B /etc/fstab** this
```
/dev/mmcblk0p1   /uboot               vfat       defaults,sync,auto    0  0
/dev/mmcblk0p3   /                    ext3       defaults              0  0
/dev/mmcblk0p4   /data                ext3       defaults              0  0
```
right? (As I do not need more than 4 partitions right now it should be ok to skip the extended one.)

**Comment by kacf** [comment]
Yep, that looks good. On master we skip the extended partition as well. This was a bug in Yocto itself, not in our code.

**Comment by fkromer** [comment]
Great, I will add the current insights into the integration procedure to [my docs fork](https://github.com/fkromer/mender-docs/blob/master/07.Generic-device-integration/03.Build-system-overview/docs.md) and merge it with [your post info](https://mender.io/blog/porting-mender-to-a-non-yocto-build-system). I am stuck with relation to some board specifics (my board != BeagleBone Black) which blocks me right now...

This partition specific configuration of the rootfs updates (artifacts) means that I need to ensure that they are not updated onto the wrong partition by mistake, right? Could you give me some hint where I can find info about that (e.g. in [Modifying a Mender Artifact](https://docs.mender.io/1.0/Artifacts/Modifying-a-Mender-Artifact))?

**Comment by kacf** [comment]
> This partition specific configuration of the rootfs updates (artifacts) means that I need to ensure that they are not updated onto the wrong partition by mistake, right? Could you give me some hint where I can find info about that (e.g. in Modifying a Mender Artifact)?

I'm not sure I understand the question. As long as the configuration file for mender, `mender.conf` contains the rootfs A and B partitions, then it will automatically select the correct one. It shouldn't be possible for it to select the wrong one, for example it will not update to the currently running partition.

**Comment by fkromer** [comment]
I mean if I do have to consider something on the server side in addition to selecting the target artifact during the creation of the deployment (to map to the corresponding rootfs partition of the artifact)?

**Comment by kacf** [comment]
Ah no, certainly not. The server doesn't even know which partition the client will use, this is all handled by the client, and it will just report back if there's a problem.

**Comment by fkromer** [comment]
Great, thanks. That was actually the reason why I asked the [fstab related question before](https://github.com/mendersoftware/mender-docs/issues/79#issuecomment-278683268).

**Comment by fkromer** [comment]
What permissions for the boot partition (`boot.vfat`, contains `uboot.env`, mounted on `/uboot`) and the u-boot environment file (`/uboot/uboot.env`) do you suggest?

**Comment by kacf** [comment]
`0644` should be appropriate. It will need to be written to by the daemon, and the content isn't sensitive to reading.

**Comment by fkromer** [comment]
The **referencing of the u-boot environment seems to be inconsistent** right now:

In `/u-boot/include/config_mender.h` I use:
```
#define CONFIG_SYS_MMC_ENV_DEV   MENDER_UBOOT_STORAGE_DEVICE
/* Use MMC partition zero to select whole user area of memory card. */
#define CONFIG_SYS_MMC_ENV_PART  0
#define CONFIG_ENV_OFFSET        MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1
#define CONFIG_ENV_OFFSET_REDUND MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_2
```

In `/u-boot/include/config_mender_defines.h` I use:

```
/* Shell variables */
#define MENDER_BOOT_PART_NUMBER 1
#define MENDER_ROOTFS_PART_A_NUMBER 2
#define MENDER_ROOTFS_PART_B_NUMBER 3
#define MENDER_UBOOT_STORAGE_INTERFACE "mmc"
#define MENDER_UBOOT_STORAGE_DEVICE 0

/* BB variables. */
#define MENDER_STORAGE_DEVICE_BASE "/dev/mmcblk0p"
#define MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1 0x800000
#define MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_2 0x1000000

/* For sanity checks. */
#define MENDER_BOOTENV_SIZE 0x20000
```

In the board specific `/u-boot/include/configs/... .h` which include each other I use:

```
#define CONFIG_ENV_IS_IN_MMC
```
```
#define CONFIG_ENV_SIZE			0x20000
```

and in `/u-boot/tools/env/fw_env.config` I use:

```
/uboot/uboot.env  0x0000  0x20000
```

The u-boot environment file `uboot.env` is located in partition `boot.vfat` (`mmcblk0p1`) and mounted under `/uboot/uboot.env`.

The boot partition is flashed with a offset `0x2000` (because the mbr.bin file is placed in front of it) with `setenv update_2 'tftpboot ${loadaddr} ${board_name}/boot.vfat && run set_blkcnt && mmc dev 0 0 && mmc write ${loadaddr} 0x2000 ${blkcnt}'`

When I boot into u-boot there are NO error messages displayed (like `*** Error - No Valid Environment Area found` or
`*** Warning - bad CRC, using default environment`). And if I enter `Apalis TK1 # printenv` in the u-boot prompt the size seems to be correct (with relation to `0x20000`) `Environment size: 6231/131067 bytes`.

When I enter `Apalis TK1 # setenv mender_uboot_test 1; saveenv;`, boot into rootfs A and enter `ubuntu@tegra-ubuntu:~$ fw_printenv` I get some output but without `mender_uboot_test=1`.

When I then enter `ubuntu@tegra-ubuntu:~$ fw_setenv mender_linux_test 1`, reboot into the u-boot prompt and entering `Apalis TK1 # printenv`.

What did I miss that the environment referencing is inconsistent?

**Comment by kacf** [comment]
Somehow two things have gotten mixed together here. The `/uboot/uboot.env` file was only used before `CONFIG_ENV_IS_IN_MMC` was introduced, and is therefore obsolete. Originally I thought you were modelling your changes after the old `CONFIG_ENV_IS_IN_FAT` method, but since you've started on the MMC variant, I recommend to continue with this.

This means that you can remove the `/uboot/uboot.env` file completely. In fact, the whole partition is unneeded unless you need to place a bootloader on it.

In `fw_env.config` you should put:

```
/dev/mmcblk0 0x800000 0x20000
/dev/mmcblk0 0x1000000 0x20000
```

And I think the rest should be correct.

**Comment by kacf** [comment]
Be aware as well, that the space used by the environment above must be accounted for when you create the partition. So there needs to be enough of a gap between the start of the device and the first partition to have room for the above. In the case above there is 8MiB alignment, so it will be 24MiB in (0x1800000).

**Comment by fkromer** [comment]
Initially I tried to follow [your posts](https://mender.io/blog/porting-mender-to-a-non-yocto-build-system) approach holding the u-boot environment in `/uboot/uboot.env` with the [docs](https://docs.mender.io/1.0/Devices/Integrating-with-U-Boot#u-boot-features) for more detailed info about it (which turned out to deviate :smile:).

Now I am thinking about which approach to follow:
Keeping the u-boot environment in a single file in `/uboot/uboot.env` without redundancy is a bit risky I guess. But I one is able to keep and to update the u-boot environment independent from the rootfs partition. In comparison keeping the u-boot environments directly in the eMMC has the advantage of redundancy. Are there other issues which one should consider with relation to this comparison?

**Comment by kacf** [comment]
That's a pretty accurate assessment. The environment was moved to redundant MMC for two reasons:

* To avoid losing the environment state if machine loses power while writing the environment file.
* To avoid corrupting the FAT partition if the machine loses power while writing the FAT allocation table.

**Comment by fkromer** [comment]
> So there needs to be enough of a gap between the start of the device and the first partition to have room for the above. In the case above there is 8MiB alignment, so it will be 24MiB in (0x1800000).

Right now I am flashing `mmcblk0` with
- `mbr.bin` to blkstart `0x0` (blkcnt of `0x1`)
- `boot.vfat` to blkstart `0x2000`
- `rootfs_a.ext3` to blkstart `0xa000`
- `rootfs_b.ext3` to blkstart `0xbc2000`
- etc.

With the MMC approach and not considering the alignment I would have to move the blkstart of `boot.vfat` to greater than `0x1+0x20000` (environment size of `0x20000`), checking that blkstart+blkcnt of `boot.vfat` is not overlapping with the blkstart of `rootfs_a.ext3` and if overlapping moving `rootfs_a.ext3` and `rootfs_b.ext3` accordingly, right? How do I have to consider the alignment?

**Comment by fkromer** [comment]
Has there something changed with [Mender client release 1.0.0](https://github.com/mendersoftware/mender/tree/1.0.0) with relation to the integration docs?

**Comment by fkromer** [comment]
I am able to set the u-boot environment variable for testing `mender_uboot_test` in u-boot with `setenv` + `saveenv` and to read the environment from within rootfs A and rootfs B with `fw_printenv`. But if I set the variable `mender_linux_test` within either rootfs A or rootfs B with `fw_setenv` I am just able to read it in the corresponding rootfs with `fw_printenv`. But I am not able to read it from within u-boot with `printenv`.

I am a bit concerned about that the [commit which changed from the u-boot environment in a file on the boot partition to the redundant environment variables in front of the root file systems](https://github.com/mendersoftware/meta-mender/commit/2a8ea346aa0d12da755d36f426164a0e54f757b3) still states something about `uboot.env` and `uEnv.txt`. Is that in there for backward compatibility reasons? Or do I have to consider something in addition to the u-boot environment in front of the root file systems instead of the boot partition file.

**Comment by fkromer** [comment]
Are you guys still interested in continuing this issue (docs about non-yocto build integration)? Or should I stick to your [Mender Software Support](https://mender.io/product/support-and-services) and continue in private discussion :wink:?

**Comment by fkromer** [comment]
With relation to the [`mender-beaglebone.sdimg.gz` from the docs](https://docs.mender.io/1.0/Getting-started/Deploy-to-physical-devices#prepare-the-disk-image) you mean I need enough space between the rootfs B partition `mender-beaglebone.sdimg3` and the data partition `mender-beaglebone.sdimg5` (`1802240 - 1785855 = 16384 = 8192*2`), right?

    kromer@rc-019:~/Downloads$ sudo fdisk -lu mender-beaglebone.sdimg
    [sudo] password for kromer: 
    
    Disk mender-beaglebone.sdimg: 1056 MB, 1056964608 bytes
    4 heads, 32 sectors/track, 16128 cylinders, total 2064384 sectors
    Units = sectors of 1 * 512 = 512 bytes
    Sector size (logical/physical): 512 bytes / 512 bytes
    I/O size (minimum/optimal): 512 bytes / 512 bytes
    Disk identifier: 0xa78c02df
    
                      Device Boot      Start         End      Blocks   Id  System
    mender-beaglebone.sdimg1   *       49152       81919       16384    c  W95 FAT32 (LBA)
    mender-beaglebone.sdimg2           81920      933887      425984   83  Linux
    mender-beaglebone.sdimg3          933888     1785855      425984   83  Linux
    mender-beaglebone.sdimg4         1802239     2064383      131072+   f  W95 Ext'd (LBA)
    mender-beaglebone.sdimg5         1802240     2064383      131072   83  Linux

**Comment by estenberg** [comment]
@fkromer we certainly want to assist with making Mender more broadly available and thanks for all the work here.

I am not an expert in this area, so I can not help much with your questions here, unfortunately, @kacf is our best bet and I think he will get back. But is your goal still to adapt a stock BeagleBone image to integrate Mender OTA? Or are you working with a sample build system?

**Comment by fkromer** [comment]
@estenberg I am still working on the improvement of the generic, Non-Yocto build system docs which allows feedback and discussion besides ones (possible) commercial/job background. The only major difference is that  I wanted to use the eMMC instead of the SD card to hold the image right from the beginning. (But this makes a major differences with relation to the development workflow only.) When I was blocked due to the rootfs size of the official BeagleBone Black image I decided to switch to my actual development board. But as I am using the linaro for arm toolchain the steps needed to get an image (including u-boot, partitions, etc.) compatible for the integration with Mender and integrating it are basically the same.

> The final option is to get a SD card for this for now, with at least 8GB size. :-)

I think I will stick to your suggestion to get this docs done.

**Comment by fkromer** [comment]
I still don't understand why the u-boot environment location is just explicitly specified for the [beaglebone u-boot source](https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-beaglebone/recipes-bsp/u-boot/patches/0001-BBB-Use-Mender-boot-code-for-selecting-boot-device-a.patch#L115) but neither for the [QEMU](https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-qemu/recipes-bsp/u-boot/patches/0001-Enable-U-Boot-on-QEMU-to-use-Mender-boot-code.patch) nor for the [rpi](https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-raspberrypi/recipes-bsp/u-boot/patches/0001-CONFIGS-rpi-enable-mender-requirements.patch). How do they know about from where to load the u-boot environment from its customized location?

**Comment by fkromer** [comment]
Is `MENDER_PARTITION_ALIGNMENT_MB` of `4 MB` ok as well or is the valid range restricted somehow?

**Comment by kacf** [comment]
Hey @fkromer, sorry for the delay, I've been out of the office for the last week!

A lot of questions here, so let me try to take one thing at a time:

- `uboot.env` vs `uEnv.txt`: I think there is some confusion here: These two files are two different and completely unrelated pieces of the U-Boot environment. The latter, `uEnv.txt` you can mostly forget about, it's only for customized boot scripts, and it has nothing to do with the redundant U-Boot environment, `setenv`/`saveenv`/`printenv` or the gap between partitions. As for the `uboot.env`, this is the file that was previously used to store the environment on the first partition. However, with the introduction of redundant U-Boot environment, this file is no longer used. There are a few references to it in meta-mender still, but this is just implementation details, it never actually ends up on the device as a file.
- Copying uboot.env to raw space: Actually, this is not needed. Both U-Boot itself and the user space tools will simply revert to their built in defaults if the environment check sum is incorrect, which it will be if you haven't written it. Then they will write it for you on the next update, you just have to leave enough space.

> I still don't understand why the u-boot environment location is just explicitly specified for the beaglebone u-boot source but neither for the QEMU nor for the rpi. How do they know about from where to load the u-boot environment from its customized location?

Following from my section about `uEnv.txt` above, this is a red herring. This is something specific they do for Beaglebone in order to control the boot sequence via a "batch file". Not related to the core environment, and not required for Mender to work.

It's unfortunate they chose such similar terminology...

> With relation to the mender-beaglebone.sdimg.gz from the docs the space between the partition table and the first partition is almost 49152s * 512 bytes/s = 25165824 bytes / 24576 KiB / 24 MiB, right?

Correct.

> Has there something changed with Mender client release 1.0.0 with relation to the integration docs?

No.

> Is MENDER_PARTITION_ALIGNMENT_MB of 4 MB ok as well or is the valid range restricted somehow?

No, it just needs to be a multiple of the sector size (we plan to change this variable to accept KB, btw).

I hope that helps! Just FYI, I will be out of the office next week as well.

**Comment by fkromer** [comment]
@kacf Thank you very much for the detailed answers.

> - uboot.env vs uEnv.txt: ...
> - Copying uboot.env to raw space: Actually, this is not needed. Both U-Boot itself and the user space tools will simply revert to their built in defaults if the environment check sum is incorrect, which it will be if you haven't written it. Then they will write it for you on the next update, you just have to leave enough space.

That were the "remaining pieces in the puzzle" for me. The messages `Warning: Bad CRC, using default environment` after first boot into u-boot (and before storing it) and when accessing the u-boot environment from user space with `fw_setenv`/`fw_printenv` the first time confused me somehow in the beginning. But as you explained it makes totally sense [due to the "not yet initialized environment"](http://www.denx.de/wiki/DULG/WarningBadCRCUsingDefaultEnvironment).

After having some trouble with relation to the board specific flashing of the eMMC I got the writing/reading of the u-boot environment working.

It would be great if you could give me another hint with relation to testing before you are out of office :wink:: The [integration checklist](https://docs.mender.io/1.0/Devices/Integration-checklist) describes how to test the u-boot-fw-utils functionality in general. But how do I test if the redundant u-boot environment is used like expected in bad case situations (or is that already included in the general u-boot-fw-utils testing)?

**Comment by fkromer** [comment]
**Sidenote:**
In the [u-boot git repo](http://git.denx.de/?p=u-boot.git;a=blob;f=README;h=4f0dbd4fca6ddc94aa26e874be54231089fc79a0;hb=HEAD) in the `CONFIG_ENV_IS_IN_MMC` section in L4157 and following (and the `CONFIG_ENV_IS_IN_FLASH` section) it is stated that  `CONFIG_ENV_SIZE_REDUND` needs to be explicitly set to the same value as `CONFIG_ENV_SIZE` in addition to `CONFIG_ENV_OFFSET_REDUND`. Would it then make sense to add another build check for that in the corresponding [section of your patch](https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-core/recipes-bsp/u-boot/patches/0001-Generic-boot-code-for-Mender.patch#L65-L71)?

**Comment by kacf** [comment]
Yes, it does. :-)

Do you want to make a patch for it?

**Comment by fkromer** [comment]
DONE ([patch](https://github.com/mendersoftware/meta-mender/pull/260))
